GCE kube-down: Delete all remaining firewall rules when DELETE_NETWORK is set

MrHohn opened the pull request:
**What this PR does / why we need it**: From https://github.com/kubernetes/kubernetes/issues/52347#issuecomment-335245693, we think it'd be reasonable to cleanup firewall resources as well during GCE kube-down.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #NONE

**Special notes for your reviewer**:
/assign @shyamjvs @bowei 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

MrHohn committed:
GCE kube-down: Delete all remaining firewall rules when KUBE_DELETE_NETWORK is set

bowei commented:
Are these leaked? There are a number of other resources as well that reference network (routes, etc).

/lgtm

bowei commented:
/lgtm cancel

bowei commented on cluster/gce/util.sh:
```diff
@@ -912,6 +912,16 @@ function detect-subnetworks() {
 }
 
+function delete-all-firewall-rules-in-network() {
+  local -r all_firewalls=`gcloud compute firewall-rules list --project "${NETWORK_PROJECT}" --filter="network=${NETWORK}" --format="value(name)"`
```
use `$(gcloud compute firewall-rules list --project "${NETWORK_PROJECT}" --filter="network=${NETWORK}" --format="value(name)")`

MrHohn replied:
Done.

bowei commented on cluster/gce/util.sh:
```diff
@@ -912,6 +912,16 @@ function detect-subnetworks() {
+function delete-all-firewall-rules-in-network() {
+  local -r all_firewalls=`gcloud compute firewall-rules list --project "${NETWORK_PROJECT}" --filter="network=${NETWORK}" --format="value(name)"`
+  if [[ -z "$all_firewalls" ]]; then
+    echo "No firewall rules left to be deleted or failed to list firewall rules."
```
This is kind of bad, it either failed OR succeeded? Can we print an error message only when there was a problem?

shyamjvs replied:
+1. Please see my suggestion in https://github.com/kubernetes/kubernetes/pull/53626/files#r143698905 which handles this.

MrHohn replied:
Fixed the logic.

shyamjvs commented on cluster/gce/util.sh:
```diff
@@ -912,6 +912,16 @@ function detect-subnetworks() {
   echo "${color_red}Could not find subnetwork with region ${REGION}, network ${NETWORK}, and project ${NETWORK_PROJECT}"
 }
 
+function delete-all-firewall-rules-in-network() {
```
I would suggest just adding the following code inside `delete-network` where currently we're just listing the firewall-rules and returning in case the network failed to delete.

Just need to add an if statement there like this:

```
if fws=$(gcloud compute firewall-rules list --project "${NETWORK_PROJECT}" --filter="network=${NETWORK}" --format="value(name)"); then
  echo "Deleting firewall rules remaining in network ${NETWORK}: ${fws}"
  delete-firewall-rules "$fws"
else
  echo "Failed to list firewall rules from the network ${NETWORK}"
fi
```

MrHohn replied:
Thanks for the suggestion, seems much better and applied. Though I still put it as a separate function, it seems more reasonable if `detect-network()` only deletes network and cleanup of other resources happens in a separate function (like `delete-subnetworks()`).

MrHohn commented:
> Are these leaked? There are a number of other resources as well that reference network (routes, etc).

This is more about cleaning up the firewalls associate with k8s service. We already have logics to cleanup resources that associate with k8s cluster itself (routes, instances, disks, etc.).

For the other resources associate with service (forwarding-rule, targetpool, healthcheck, etc.), they are not referencing the network so are less concerned. I'd like to get this in first and see if we will need a deeper cleanup later.

shyamjvs commented on cluster/gce/util.sh:
```diff
@@ -912,6 +912,19 @@ function detect-subnetworks() {
 
+function delete-all-firewall-rules-in-network() {
+  if fws=$(gcloud compute firewall-rules list --project "${NETWORK_PROJECT}" --filter="network=${NETWORK}" --format="value(name)"); then
+    if [[ -z "${fws}" ]]; then
```
I think you don't need this if condition. If it is empty string, you'll just pass an empty argument to `delete-firewall-rules()` which will not delete anything inside the for loop.

MrHohn replied:
Yep, I was thinking it may be more clear to call it out if there is not firewalls needed to be deleted. But SGTM to remove this if condition.

shyamjvs commented on cluster/gce/util.sh:
```diff
@@ -912,6 +912,19 @@ function detect-subnetworks() {
   echo "${color_red}Could not find subnetwork with region ${REGION}, network ${NETWORK}, and project ${NETWORK_PROJECT}"
 }
 
+function delete-all-firewall-rules-in-network() {
```
Maybe name this `delete-all-firewall-rules`? `in-network` seems unnecessary also given that it's not actually taking the network as an argument.

MrHohn replied:
Sounds good, done.

shyamjvs commented on cluster/gce/util.sh:
```diff
@@ -912,6 +912,19 @@ function detect-subnetworks() {
+      echo "No firewall rules left to be deleted."
+    else
+      echo "Deleting firewall rules remaining in network ${NETWORK}: ${fws}"
+      delete-firewall-rules "$fws"
```
Are you sure this works as intended? i.e "$fws" is passed as a list of arguments to `delete-firewall-rules` instead of a single space-separated string?

MrHohn replied:
Yep, it worked on my test script, let me rerun it to confirm.

shyamjvs commented:
Few comments and lgtm. Also can you verify if this works (by creating some non-default firewall and then kube-down'ing or sth)?

shyamjvs commented on cluster/gce/util.sh:
```diff
@@ -1728,6 +1737,8 @@ function kube-down() {
 
     if [[ "${KUBE_DELETE_NETWORK}" == "true" ]]; then
+      # Delete all remaining firewall rules in the network.
+      delete-all-firewall-rules || true
```
Maybe call this inside `delete-network` and get rid of the `# might fail if there are leaked firewall rules` comment below?

MrHohn replied:
Ref https://github.com/kubernetes/kubernetes/pull/53626#discussion_r143827108, WDYT about keeping resource cleanup detached?

I revised that comment a bit as it isn't completely right. Network deletion might fail no only due to firewalls, but also routes, VMs, or any others that reference this network.

shyamjvs replied:
Sounds good.

MrHohn commented:
> Few comments and lgtm. Also can you verify if this works (by creating some non-default firewall and then kube-down'ing or sth)?

Seems it works, logs from my local run:
```
Bringing down cluster
Deleting firewall rules remaining in network e2e-test-*: k8s-test-1
k8s-test-2
test-delete-1
Deleted [https://www.googleapis.com/compute/v1/projects/*/global/firewalls/k8s-test-1].
Deleted [https://www.googleapis.com/compute/v1/projects/*/global/firewalls/k8s-test-2].
Deleted [https://www.googleapis.com/compute/v1/projects/*/global/firewalls/test-delete-1].
Deleted [https://www.googleapis.com/compute/v1/projects/*/global/networks/e2e-test-*].
...
```

shyamjvs commented:
/lgtm
Thanks for the fix!

bowei commented:
/lgtm

bowei commented:
/approve no-issue
